### ISSUE TITLE
Can't use database schema with a dash in its name (liquibase, postgres)

### ISSUE THREAD

**Issue by dasFlerchi** [keycloak/keycloak#20870]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

storage

### Describe the bug

I created a clean lab environment with keycloak v21.1.1 and postgres 12, both via docker container. The database was prepared with dbname `sso-stack-lab`, schema `keycloak-lab` and a user `keycloak-lab`. Configfile `keycloak.conf` is mounted from local host into the container, setting `db=postgres`, `db-url-host=localhost`, `db-username=keycloak-lab`, `db-password=mySecretPWD`, `db-schema=keycloak-lab`

When starting keycloak (from a custom image) I get the error `liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"`

After stumbling across #17276 I changed the database name to no avail. Next I tried changing the schema name to `keycloaklab` (removing the dash) and now it works. I wasnt able to further pinpoint the problem, but after removing the dash from schema name, but with username name and database name still having a dash, it works.

### Version

21.1.1

### Expected behavior

Using a database schema name with a dash, keycloak starts and works as expected.

### Actual behavior

When starting keycloak (from a custom image with added certificates and build run done) with a database schema name including a dash I get the error `liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"` and keycloak will fail to start.

### How to Reproduce?

Run a clean install of keycloak, using postgres with a database schema name including a dash.

**Comment by hmlnarik** [comment]
@dasFlerchi Could you please share the full exception?

**Comment by dasFlerchi** [comment]
> @dasFlerchi Could you please share the full exception?

Will try to get my hands on that system over the weekend and share the corresponding logs.

**Comment by dasFlerchi** [comment]
@hmlnarik 
I replicated the setup on my box at home using wsl, getting the same behaviour as on my lab setup at work.
Configuration was set in an identical way via keycloak.conf, this time using `db-username=sso-dbuser`, `db-schema=sso-schemaname`, `db-url-database=sso-dbname`. Changing the schema to `ssoschemaname` (removing the dash) solved the problem. Reverting back to `sso-schemaname` triggered it again.

Looks like the schema name within the SQL is not getting quoted, which is required.

Corresponding loglines:
```
sso-stack-bugtest-keycloak-1  | 2023-06-10 11:41:50,872 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
sso-stack-bugtest-keycloak-1  | 2023-06-10 11:41:50,872 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: liquibase.exception.LiquibaseException: liquibase.exception.UnexpectedLiquibaseException: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
sso-stack-bugtest-keycloak-1  |   Position: 23 [Failed SQL: (0) SET SEARCH_PATH TO sso-schemaname, "$user","public"]
sso-stack-bugtest-keycloak-1  | 2023-06-10 11:41:50,873 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: liquibase.exception.UnexpectedLiquibaseException: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
sso-stack-bugtest-keycloak-1  |   Position: 23 [Failed SQL: (0) SET SEARCH_PATH TO sso-schemaname, "$user","public"]
sso-stack-bugtest-keycloak-1  | 2023-06-10 11:41:50,873 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
sso-stack-bugtest-keycloak-1  |   Position: 23 [Failed SQL: (0) SET SEARCH_PATH TO sso-schemaname, "$user","public"]
sso-stack-bugtest-keycloak-1  | 2023-06-10 11:41:50,873 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: ERROR: syntax error at or near "-"
sso-stack-bugtest-keycloak-1  |   Position: 23 [Failed SQL: (0) SET SEARCH_PATH TO sso-schemaname, "$user","public"]
sso-stack-bugtest-keycloak-1  | 2023-06-10 11:41:50,874 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: ERROR: syntax error at or near "-"
sso-stack-bugtest-keycloak-1  |   Position: 23
```

**Comment by vramik** [comment]
I was able to reproduce the issue. There is full stacktrace.
```
java.lang.RuntimeException: liquibase.exception.LiquibaseException: liquibase.exception.UnexpectedLiquibaseException: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
  Position: 22 [Failed SQL: (0) SET SEARCH_PATH TO kc-db, "$user","public"]
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.lambda$getLiquibaseUnrunChangeSets$0(QuarkusJpaUpdaterProvider.java:299)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1220)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.getLiquibaseUnrunChangeSets(QuarkusJpaUpdaterProvider.java:295)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.validateChangeSet(QuarkusJpaUpdaterProvider.java:265)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.validate(QuarkusJpaUpdaterProvider.java:237)
	at org.keycloak.quarkus.runtime.storage.legacy.database.LegacyJpaConnectionProviderFactory.createOrUpdateSchema(LegacyJpaConnectionProviderFactory.java:255)
	at org.keycloak.quarkus.runtime.storage.legacy.database.LegacyJpaConnectionProviderFactory.postInit(LegacyJpaConnectionProviderFactory.java:129)
	at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.init(QuarkusKeycloakSessionFactory.java:105)
	at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.createSessionFactory(QuarkusKeycloakApplication.java:45)
	at org.keycloak.services.resources.KeycloakApplication.startup(KeycloakApplication.java:125)
	at org.keycloak.quarkus.runtime.integration.QuarkusLifecycleObserver.onStartupEvent(QuarkusLifecycleObserver.java:37)
	at org.keycloak.quarkus.runtime.integration.QuarkusLifecycleObserver_Observer_onStartupEvent_c9888fa6c2aa9208d4625ee9d83de6fd77e22c83.notify(Unknown Source)
	at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:346)
	at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:328)
	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:82)
	at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:155)
	at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:106)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:111)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
	at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:98)
	at org.keycloak.quarkus.runtime.cli.command.AbstractStartCommand.run(AbstractStartCommand.java:37)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:100)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: liquibase.exception.LiquibaseException: liquibase.exception.UnexpectedLiquibaseException: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
  Position: 22 [Failed SQL: (0) SET SEARCH_PATH TO kc-db, "$user","public"]
	at liquibase.Liquibase.runInScope(Liquibase.java:1623)
	at liquibase.Liquibase.listUnrunChangeSets(Liquibase.java:1279)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.lambda$getLiquibaseUnrunChangeSets$0(QuarkusJpaUpdaterProvider.java:297)
	... 40 more
Caused by: liquibase.exception.UnexpectedLiquibaseException: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
  Position: 22 [Failed SQL: (0) SET SEARCH_PATH TO kc-db, "$user","public"]
	at liquibase.changelog.StandardChangeLogHistoryService.hasDatabaseChangeLogTable(StandardChangeLogHistoryService.java:88)
	at liquibase.changelog.StandardChangeLogHistoryService.getRanChangeSets(StandardChangeLogHistoryService.java:317)
	at org.keycloak.connections.jpa.updater.liquibase.conn.CustomChangeLogHistoryService.getRanChangeSets(CustomChangeLogHistoryService.java:54)
	at liquibase.database.AbstractJdbcDatabase.getRanChangeSetList(AbstractJdbcDatabase.java:1156)
	at liquibase.changelog.DatabaseChangeLog.validate(DatabaseChangeLog.java:342)
	at liquibase.Liquibase.lambda$listUnrunChangeSets$16(Liquibase.java:1287)
	at liquibase.Scope.lambda$child$0(Scope.java:194)
	at liquibase.Scope.child(Scope.java:203)
	at liquibase.Scope.child(Scope.java:193)
	at liquibase.Scope.child(Scope.java:172)
	at liquibase.Liquibase.runInScope(Liquibase.java:1618)
	... 42 more
Caused by: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
  Position: 22 [Failed SQL: (0) SET SEARCH_PATH TO kc-db, "$user","public"]
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:445)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:77)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:160)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:125)
	at liquibase.database.core.DatabaseUtils.initializeDatabase(DatabaseUtils.java:61)
	at liquibase.database.core.PostgresDatabase.rollback(PostgresDatabase.java:403)
	at liquibase.snapshot.SnapshotGeneratorFactory.has(SnapshotGeneratorFactory.java:118)
	at liquibase.snapshot.SnapshotGeneratorFactory.hasDatabaseChangeLogTable(SnapshotGeneratorFactory.java:269)
	at liquibase.changelog.StandardChangeLogHistoryService.hasDatabaseChangeLogTable(StandardChangeLogHistoryService.java:86)
	... 52 more
Caused by: org.postgresql.util.PSQLException: ERROR: syntax error at or near "-"
  Position: 22
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2713)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2401)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:368)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:498)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:415)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:335)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:321)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:297)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:292)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.postgresql.ds.PGPooledConnection$StatementHandler.invoke(PGPooledConnection.java:441)
	at jdk.proxy2/jdk.proxy2.$Proxy75.execute(Unknown Source)
	at io.agroal.pool.wrapper.StatementWrapper.execute(StatementWrapper.java:235)
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:441)
	... 60 more
```

**Comment by hmlnarik** [comment]
Thank you @vramik. Caused by https://github.com/liquibase/liquibase/issues/1640

Can we work around it anyhow?

**Comment by vramik** [comment]
I was testing a workaround suggested in the [original issue](https://github.com/liquibase/liquibase/issues/3205#issuecomment-1377621909) and I can confirm keycloak starts correctly when  [`LIQUIBASE_PRESERVE_SCHEMA_CASE`](https://docs.liquibase.com/parameters/home.html) set to `true`:

```bash
[vramik@localhost tmp]$ export LIQUIBASE_PRESERVE_SCHEMA_CASE=true
[vramik@localhost tmp]$ echo $LIQUIBASE_PRESERVE_SCHEMA_CASE
true
[vramik@localhost tmp]$ ./keycloak-999.0.0-SNAPSHOT/bin/kc.sh start-dev --debug --db postgres --db-url=jdbc:postgresql://localhost:5432/keycloak --db-username=keycloak --db-password=pass --db-schema="kc-db"
Listening for transport dt_socket at address: 8787
Listening for transport dt_socket at address: 8787
2023-06-13 10:49:26,394 INFO  [org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (main) Hostname settings: Base URL: <unset>, Hostname: <request>, Strict HTTPS: false, Path: <request>, Strict BackChannel: false, Admin URL: <unset>, Admin: <request>, Port: -1, Proxied: false
2023-06-13 10:49:27,507 WARN  [io.quarkus.agroal.runtime.DataSources] (main) Datasource <default> enables XA but transaction recovery is not enabled. Please enable transaction recovery by setting quarkus.transaction-manager.enable-recovery=true, otherwise data may be lost if the application is terminated abruptly
2023-06-13 10:49:27,905 WARN  [org.infinispan.PERSISTENCE] (keycloak-cache-init) ISPN000554: jboss-marshalling is deprecated and planned for removal
2023-06-13 10:49:28,000 INFO  [org.infinispan.CONTAINER] (keycloak-cache-init) ISPN000556: Starting user marshaller 'org.infinispan.jboss.marshalling.core.JBossUserMarshaller'
2023-06-13 10:49:28,343 INFO  [org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory] (main) Node name: node_625142, Site name: null
2023-06-13 10:49:29,287 INFO  [org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider] (main) Initializing database schema. Using changelog META-INF/jpa-changelog-master.xml

UPDATE SUMMARY
Run:                        114
Previously run:               0
Filtered out:                 0
-------------------------------
Total change sets:          114

2023-06-13 10:49:33,396 INFO  [org.keycloak.broker.provider.AbstractIdentityProviderMapper] (main) Registering class org.keycloak.broker.provider.mappersync.ConfigSyncEventListener
2023-06-13 10:49:33,411 INFO  [org.keycloak.services] (main) KC-SERVICES0050: Initializing master realm
2023-06-13 10:49:34,985 INFO  [io.quarkus] (main) Keycloak 999.0.0-SNAPSHOT on JVM (powered by Quarkus 3.1.0.Final) started in 9.463s. Listening on: http://0.0.0.0:8080
2023-06-13 10:49:34,986 INFO  [io.quarkus] (main) Profile dev activated. 
2023-06-13 10:49:34,986 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, jdbc-h2, jdbc-mariadb, jdbc-mssql, jdbc-mysql, jdbc-oracle, jdbc-postgresql, keycloak, logging-gelf, micrometer, narayana-jta, reactive-routes, resteasy, resteasy-jackson, smallrye-context-propagation, smallrye-health, vertx]
2023-06-13 10:49:34,989 WARN  [org.keycloak.quarkus.runtime.KeycloakMain] (main) Running the server in development mode. DO NOT use this configuration in production.

```

**Comment by vramik** [comment]
The issue is blocked by original issue in liquibase. There is no plan to make any changes to keycloak's codebase regarding it, waiting for bumping liquibase version containing the fix.

**Comment by ahus1** [comment]
@vramik - this issue came up in a different context from another context. When reading the parent issue, I see that adding the option in all cases this might cause some regression issues. 

So I wonder if we could have a logic like 

> If the db schema name contains a dash, call `System.setProperty(GlobalConfiguration.PRESERVE_SCHEMA_CASE.getKey(), "true");`"

until the upstream issue is fixed.

**Comment by vramik** [comment]
Thank you @ahus1,

> So I wonder if we could have a logic like
> 
> > If the db schema name contains a dash, call `System.setProperty(GlobalConfiguration.PRESERVE_SCHEMA_CASE.getKey(), "true");`"
> 
> until the upstream issue is fixed.

It would be possible to have such a logic. Just note that for a (minor) subset of users who rely on case insensitive schema name and have a dash in it the workaround wouldn't work. 

Could you please elaborate a bit on "a different context from another context"?

**Comment by ahus1** [comment]
As discussed today between @vramik and me: It would make it simpler for customers if Keycloak would include the workaround whenever the DB schema contains a dash (`-`). The current error message is not helpful, and doesn't reveal the cause nor the workaround/remedy.

It should print a warning when doing so. We'll track this in a follow-up issue to remove the workaround as when a fix in Liquibase is available.

**Comment by ahus1** [comment]
The next release of Keycloak will contain the described workaround. Unfortunately this workaround will only work for PostgreSQL due to the code in Liquibase.

The follow-up issue to track this: #23174

**Comment by hungrymonkey** [comment]
My apologies for pinging this issue.
